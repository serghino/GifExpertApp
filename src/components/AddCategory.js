import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    // this state should contain a value by default.
    const [inputValue, setInputValue] = useState(''); // ''

    const handleInputChange = ( e ) => {
        setInputValue( e.target.value );
    }

    const handleSubmit = (e) => {
        // prevent auto refresh of the main page.
        e.preventDefault();

        if ( inputValue.trim().length > 2 ) {
            setCategories( cats => [ inputValue, ...cats, ] );
            setInputValue('');
        }

    }

    return (
        <form onSubmit={ handleSubmit }>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
