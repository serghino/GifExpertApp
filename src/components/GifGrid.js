import React from 'react';
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const { data:images, loading, total } = useFetchGifs( category );

    return (
        <>
            <div className="gridTitle">
                <h3 className="animate__animated animate__fadeIn"> { category } </h3>
                <p><b>Total</b>: {total}</p>
            </div>

            { loading && <p className="animate__animated animate__flash">Loading</p> }

            <div className="card-grid">
                
                {
                    images.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }
            
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}