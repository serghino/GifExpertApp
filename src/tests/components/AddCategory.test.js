import React from "react";
import '@testing-library/jest-dom';
import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe("Test component <AddCategory />", () => {
  const setCategories = jest.fn();
  // const setCategories = () => {};
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);
  
  beforeEach(()=> {
      jest.clearAllMocks();
      wrapper = shallow(<AddCategory setCategories={setCategories} />);
    });

  test("should show <GifGridItem /> correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should change input text", () => {
    const input = wrapper.find("input");
    const value = "hello world";
    // simulate event change event
    input.simulate("change", { target: { value } });
    // after the value is injected.
    const p = wrapper.find('p');
    expect(p.text().trim()).toBe( value );
  });
  test("should not send the form", () => {
    wrapper.find('form').simulate('submit',{preventDefault(){}});
    expect(setCategories).not.toHaveBeenCalled();
  });
  test("should send the form and clean the input text", () => {
    const value = 'simpsons';
    // simulate event change event
    wrapper.find("input").simulate("change", { target: { value } });
    // simulate submit event
    wrapper.find('form').simulate('submit',{preventDefault(){}});
    // expected
    expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledTimes(1);
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
    expect(wrapper.find("input").prop('value')).toBe('');
  });
});
