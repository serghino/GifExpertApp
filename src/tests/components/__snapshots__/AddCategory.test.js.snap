// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component <AddCategory /> should show <GifGridItem /> correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
     
     
  </p>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
